### PR TITLE
Limit to-device EDU sizes

### LINKED ISSUE #17035: Strange federation bug. Possibly in Synapse , matrix.org or both.
### Description

Hi

This will be a vague bug report so I'm hoping someone will come with a "aha" moment when reading this.

For the past week my homeserver has been unable to send outgoing federation messages towards matrix.org. Initially after starting synapse some messages go through but after a few seconds it stops and goes into a retry loop. 
During these retries matrix.org , or rather cloudflare, returns a http error of 520.

Upon further inspection i managed to narrow it down to a m.direct_to_device (very large json object) being posted by a single user.
After deleting the event from device_federation_outbox everything worked again.

My theory is that the object was too large so matrix.org/cloudflare threw an error and Synapse just kept retrying. 

If this is correct is there a bug in Synapse where it should somehow split this into multiple requests?
Or is it a matrix.org bug that has a low limit on requests size?

Attaching the deleted event.

Thank you
[bad_edu.log](https://github.com/element-hq/synapse/files/14774345/bad_edu.log)

### Steps to reproduce

-

### Homeserver

xmr.se

### Synapse Version

Multiple tested, 1.99 and up. Now 1.103.0

### Installation Method

pip (from PyPI)

### Database

postgresql. Single, no, no

### Workers

Multiple workers

### Platform

Not relevant.

### Configuration

_No response_

### Relevant log output

```shell
2024-03-25 19:22:26,936 - synapse.http.matrixfederationclient - 755 - INFO - federation_transaction_transmission_loop-24 - {PUT-O-29} [matrix.org] Got response headers: 520
2024-03-25 19:22:26,936 - synapse.http.matrixfederationclient - 798 - INFO - federation_transaction_transmission_loop-24 - {PUT-O-29} [matrix.org] Request failed: PUT matrix-federation://matrix.org/_matrix/federation/v1/send/1711389457952: HttpResponseException('520: ')
```

### Anything else that would be useful to know?

_No response_

### PR DESCRIPTION
**Pull request by erikjohnston**
This is based on https://github.com/element-hq/synapse/pull/18416, which got reverted (#19614) due to it incorrectly rejecting to-device messages to users with many devices (and thus breaking message sending).

Fix https://github.com/element-hq/synapse/issues/17035

A to-device message content looks like:

```jsonc
{
  "@user:domain": {"device1": {...}, "device2": {...}},
  ...
}
```

The previous PR would split up into multiple EDUs, each with a subset of the users. However, if one user's entry was too large it would not further split it up and then error out.

The main change in this PR is to allow splitting up a single user into multiple EDUs.

Other changes:
1. Rename to `SOFT_MAX_EDU_SIZE` to indicate that we sometimes send EDUs with larger size than that, and its more a target than a hard limit.
2. Check early if any to-device message (to a specific device) is too large to send, even if we're not going to send it over federation. This ensures that we catch issues where clients try to send too large to-device.

This still means that if a client send a large individual to-device message it will fail, but I don't believe we ever send such large to-device messages (normally they're in the range of a few KB).

---

I ended up changing the implementation a bunch to make it easy to reuse the code to split up dictionaries. Instead of repeatedly splitting up the EDU until each bit fits into the size, we instead record the size of each entry in the dict and instead split up based on cumulative size. This means we call `encode_canonical_json` on each entry rather than once on the entire struct, but its not significantly slower to do so.

--

cc @MatMaul @MadLittleMods

### CODE REVIEW

File: synapse/handlers/devicemessage.py [hunk 1]
```diff
@@ -436,57 +449,61 @@ def split_device_messages_into_edus(
...
+            # This message doesn't fit in a single EDU. We split the message up
+            # further by device.
+            #
+            # Note: `subset` should only have a single entry in it.
+            for recipient, messages_by_device in subset_messages.items():
+                # Account for the additional nesting and JSON encoding of the
+                # recipient.
+                # The +7 is for the quotes around the recipient, the
+                # colon and two pairs of curly braces.
+                base_size = base_edu_size + len(recipient) + 7
```
Comment (MadLittleMods) [inline synapse/handlers/devicemessage.py]: Can we use `encode_canonical_json(edu)` to avoid the magic values?

We're basically re-introducing the manual calculations we wanted to avoid https://github.com/element-hq/synapse/pull/18416#discussion_r2104165684

(also applies below)

File: changelog.d/19617.bugfix [hunk 2]
```diff
@@ -0,0 +1 @@
+A long queue of to-device messages could prevent outgoing federation because of the size of the transaction, let's limit the to-device EDU size to a reasonable value.
```
Comment (MadLittleMods) [inline changelog.d/19617.bugfix:1]: > 1. Rename to `SOFT_MAX_EDU_SIZE` to indicate that we sometimes send EDUs with larger size than that, and its more a target than a hard limit.

It seems like this makes the situation for https://github.com/element-hq/synapse/issues/17035 better but won't completely solve it?

Do we need some sort of hard limit? Like `MAX_EDUS_PER_TRANSACTION * SOFT_MAX_EDU_SIZE`

Comment (erikjohnston) [reply changelog.d/19617.bugfix:1]: Yeah, so that is the main question left. There are two angles:
1. We should assert that all to-device messages from the client are below whatever max we decide on is (whether that be `SOFT_MAX_EDU_SIZE` or larger). This should be safe enough as all crypto to-device messages should be less than `SOFT_MAX_EDU_SIZE` anyway.
2. If we choose a higher max, we need to look at the EDU sending code to make sure it can handle larger to-device messages than `SOFT_MAX_EDU_SIZE`. I think the current code always packs in multiple different types of EDU rather than gracefully handling one being very large.

Comment (erikjohnston) [reply changelog.d/19617.bugfix:1]: Talking to folks, I think we should add a hard limit to to-device content size. This will mean that technically the EDUs could be slightly bigger (due to EDU headers), but that shouldn't matter.

Comment (erikjohnston) [reply changelog.d/19617.bugfix:1]: Ah, so this is already in the PR https://github.com/element-hq/synapse/pull/19617/changes#diff-64fd6daf42afff49a445cf22515c03602ffa23720618e1cb7bfd4673db363f33R245-R250

Comment (erikjohnston) [reply changelog.d/19617.bugfix:1]: As for the renaming: given we can still send *other* EDU types without checking the size, and we may receive EDUs of larger size, I'm inclined to keep it as a `SOFT_MAX_`

File: synapse/handlers/devicemessage.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/handlers/devicemessage.py]: [7afc60e](https://github.com/element-hq/synapse/pull/19617/commits/7afc60e315b14d7182c09721e45f632a7c503352)

File: synapse/handlers/devicemessage.py [hunk 3]
```diff
@@ -397,3 +423,131 @@ async def get_events_for_dehydrated_device(
...
+    Args:
+        sender_user_id: The user that is sending the to-device messages.
+        message_type: The type of to-device messages that are being sent.
+        messages_by_user_then_device: Dictionary of recipient user_id to recipient device_id to message.
+
+    Returns:
+        Bin-packed list of EDU JSON content for the given to_device messages
+
+    Raises:
+        EventSizeError: If a single to-device message is too large to fit into an EDU.
```
Comment (MadLittleMods) [inline synapse/handlers/devicemessage.py]: I don't see `EventSizeError` raised in this code anymore

Feels like it should if we have some sort of hard-limit on a single EDU.

File: synapse/handlers/devicemessage.py [hunk 4]
```diff
@@ -233,6 +242,18 @@ async def send_device_message(
 
             # add an opentracing log entry for each message
             for device_id, message_content in by_device.items():
+                # Check the size of each message, as if these are too large we
+                # can't send them over federation.
+                #
+                # We do this for all to-device messages, even those that aren't
+                # over federation, so as to more easily catch clients that are
+                # sending excessively large messages.
```
Comment (MadLittleMods) [inline synapse/handlers/devicemessage.py:251]: We're losing some of the context from the [previous PR](https://github.com/element-hq/synapse/pull/18416), https://github.com/tchapgouv/synapse/blob/4b7da00420945815e19f400c350b3f4e58f24cb4/synapse/handlers/devicemessage.py#L474-L481

Comment (erikjohnston) [reply synapse/handlers/devicemessage.py:251]: 32134a5e640fc8b6aa05e1d7dfabde8d20095466

File: synapse/handlers/devicemessage.py (hunk 3 above)
Comment (erikjohnston) [reply synapse/handlers/devicemessage.py]: Have removed. This function won't try and limit things, as we already have applied the limits above. I think that is the best approach, especially if we want to reuse for other EDUs.

File: synapse/handlers/devicemessage.py [hunk 5]
```diff
@@ -233,6 +242,22 @@ async def send_device_message(
...
+                if len(encode_canonical_json(message_content)) > SOFT_MAX_EDU_SIZE:
+                    # 413 is currently an unspecced response for `/sendToDevice` but is
+                    # probably the best thing we can do.
+                    # https://github.com/matrix-org/matrix-spec/pull/2340 tracks adding
+                    # this to the spec
+                    raise EventSizeError(
+                        f"To-device message for {user_id}:{device_id} is too large to send",
```
Comment (anoadragon453) [inline synapse/handlers/devicemessage.py]: Could we add the computed message size to the log line to make this slightly more useful for debugging?

File: synapse/util/__init__.py [hunk 6]
```diff
@@ -165,3 +169,107 @@ def clear(self) -> None:
...
+    size of that structure. The default assumes no wrapping, and just accounts
+    for the two curly braces of the dict itself.
+
+    Note that if an individual entry in the original dict is larger than
+    `soft_max_size` then this will emit a dict containing just that entry, which
+    will be larger than `soft_max_size` when encoded as JSON.
+
+    Args:
+        original_dict: The dict to split. soft_max_size: The maximum size of
+            each dict when encoded as JSON.
```
Comment (anoadragon453) [inline synapse/util/__init__.py]:
```suggestion
        original_dict: The dict to split.
        soft_max_size: The maximum size of each dict when encoded as JSON.
```

File: synapse/util/__init__.py [hunk 7]
```diff
@@ -165,3 +169,107 @@ def clear(self) -> None:
...
+    """Estimated size of the JSON encoding of the current payload, including any
+    wrapping structure."""
+
+
+def split_dict_to_fit_to_size(
+    original_dict: dict[str, Any],
+    *,
+    soft_max_size: int,
+    wrapping_object_size: int = 2,
+) -> Iterator[tuple[dict[str, JsonDict], int]]:
```
Comment (anoadragon453) [inline synapse/util/__init__.py:192]: We should note that this is a generator in the docstring.

File: synapse/util/__init__.py [hunk 8]
```diff
@@ -165,3 +169,107 @@ def clear(self) -> None:
...
+    # The current payload being built up. We keep track of the estimated size of
+    # the JSON encoding of this payload so that we can decide when to start a
+    # new one.
+    current_payload = _DictSplitterState(subset={}, estimated_size=wrapping_object_size)
+
+    for key, payload in original_dict.items():
+        current_payload.subset[key] = payload
+        current_size = _len_with_wrapping_object(
+            current_payload.subset, wrapping_object_size
+        )
```
[the remark this answers is not in the document]
Comment (anoadragon453) [reply synapse/util/__init__.py:237]: The PR description mentions calling `encode_canonical_json` on each entry. But Copilot is correct; we are technically calling `encode_canonical_json` multiple times on each entry in `current_payload.subset` as we build it up.

I presume you'd have to store the length for each entry in subset to avoid this.

File: synapse/handlers/devicemessage.py [hunk 9]
```diff
@@ -397,3 +427,128 @@ async def get_events_for_dehydrated_device(
...
+
+    # First split up the top-level dict of user_id to device messages.
+    for subset_messages, estimated_size in split_dict_to_fit_to_size(
+        messages_by_user_then_device,
+        soft_max_size=SOFT_MAX_EDU_SIZE,
+        wrapping_object_size=base_edu_size,
+    ):
+        # The returned subset might be larger than the soft max size if it
+        # contains a single entry that is larger than the soft max size.
+        if estimated_size <= SOFT_MAX_EDU_SIZE - base_edu_size:
```
[the remark this answers is not in the document]
Comment (anoadragon453) [reply synapse/handlers/devicemessage.py]: I suspect you would need a unit test which tries to send a message that's *just* under `SOFT_MAX_EDU_SIZE` in order to stress this edge case.

File: tests/util/test_split_dict.py [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+
+from synapse.util import split_dict_to_fit_to_size
+
+from tests.unittest import TestCase
+
+logger = logging.getLogger(__name__)
+
+T = TypeVar("T")
+
+
```
[the remark this answers is not in the document]
Comment (anoadragon453) [reply tests/util/test_split_dict.py]: And yet all our linting passed 🤔

**Review by anoadragon453 — APPROVED** [review]
Overall looks great. Just some minor things below.

File: synapse/util/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/util/__init__.py:237]: That was the original implementation, but you have to be careful of handling extra chars like commas, etc.

The common case should be that the to-device content is smaller than max, and therefore we take the fast path of only serializing once.

File: tests/util/test_split_dict.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (erikjohnston) [reply tests/util/test_split_dict.py]: 6d8ce6359c

File: synapse/handlers/devicemessage.py [hunk 11]
```diff
@@ -397,3 +427,128 @@ async def get_events_for_dehydrated_device(
...
+def split_device_messages_into_edus(
+    sender_user_id: str,
+    message_type: str,
+    messages_by_user_then_device: dict[str, dict[str, JsonDict]],
+) -> list[JsonDict]:
+    """
+    This function takes many to-device messages and fits/splits them into several EDUs
+    as necessary. We split the messages up as the overall request can overrun the
+    `max_request_body_size` and prevent outbound federation traffic because of the size
+    of the transaction (cf. `MAX_EDU_SIZE`).
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/handlers/devicemessage.py]: f87d95ac6f

File: synapse/handlers/devicemessage.py (hunk 5 above)
Comment (erikjohnston) [reply synapse/handlers/devicemessage.py]: d2baa574d8

File: synapse/util/__init__.py (hunk 6 above)
Comment (erikjohnston) [reply synapse/util/__init__.py]: 65011eb2c7

File: synapse/handlers/devicemessage.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/handlers/devicemessage.py]: copilot wasn't wrong, fixed and added a test in 9f9935bae3

File: synapse/util/__init__.py (hunk 7 above)
Comment (erikjohnston) [reply synapse/util/__init__.py:192]: We could annotate that it's a generator, but the fact its a generator is an implementation detail? A generator like this *is* an `Iterator`

**Comment by MatMaul** [comment]
Thank you all for getting this through :heart:

File: tests/rest/client/test_sendtodevice.py [hunk 12]
```diff
@@ -91,6 +122,238 @@ def test_user_to_user(self) -> None:
...
+
+        self.get_success(sender.send_device_messages([destination]))
+
+        number_of_edus_sent = 0
+        for call in mock_send_transaction.call_args_list:
+            number_of_edus_sent += len(call[0][1]()["edus"])
+
+        self.assertEqual(number_of_edus_sent, 2)
+
+    def test_edu_large_messages_not_splitting_one_user(self) -> None:
```
Comment (MadLittleMods) [inline tests/rest/client/test_sendtodevice.py:219]: Just saw this test flake in another PR,

:x: https://github.com/element-hq/synapse/actions/runs/26910432685/job/79387059438?pr=19648
```
Error: 
Traceback (most recent call last):
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.14/lib/python3.14/site-packages/twisted/logger/_observer.py", line 81, in __call__
    observer(event)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.14/lib/python3.14/site-packages/twisted/logger/_legacy.py", line 90, in __call__
    self.legacyObserver(event)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.14/lib/python3.14/site-packages/twisted/trial/_dist/workertrial.py", line 44, in emit
    self.protocol.callRemote(managercommands.TestWrite, out=text)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.14/lib/python3.14/site-packages/twisted/protocols/amp.py", line 951, in callRemote
    return co._doCommand(self)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.14/lib/python3.14/site-packages/twisted/protocols/amp.py", line 1973, in _doCommand
    d = proto._sendBoxCommand(
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.14/lib/python3.14/site-packages/twisted/protocols/amp.py", line 884, in _sendBoxCommand
    box._sendTo(self.boxSender)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.14/lib/python3.14/site-packages/twisted/protocols/amp.py", line 713, in _sendTo
    proto.sendBox(self)
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.14/lib/python3.14/site-packages/twisted/protocols/amp.py", line 2358, in sendBox
    self.transport.write(box.serialize())
  File "/home/runner/.cache/pypoetry/virtualenvs/matrix-synapse-pswDeSvb-py3.14/lib/python3.14/site-packages/twisted/protocols/amp.py", line 692, in serialize
    raise TooLong(False, True, v, k)
twisted.protocols.amp.TooLong: 

tests.rest.client.test_sendtodevice.SendToDeviceTestCase.test_edu_large_messages_not_splitting_one_user
```

Previous discussions in the original PR:

 - https://github.com/element-hq/synapse/pull/18416#discussion_r2289444806
 - https://github.com/element-hq/synapse/pull/18416#discussion_r2662543891
 
How we worked around this before:

 - https://github.com/element-hq/synapse/pull/18736
